web_external/util/annotationGeometryParser.js

```diff
@@ -1,154 +1,261 @@
         return this.copy();
     }
 
+    //-------------------------------------------------------------------------
     changeTrack(trackId, newTrackId) {
         if (trackId !== newTrackId) {
-            this._trackIds.delete(trackId);
-            this._trackIds.add(newTrackId);
-            this._trackRanges.set(newTrackId, this._trackRanges.get(trackId));
-            this._enableState.set(newTrackId, this._enableState.get(trackId));
-            _.flatten(Array.from(this._frameMap.values())).forEach((geom) => {
-                if (geom.id1 !== trackId) {
-                    return;
-                }
-                geom.id1 = newTrackId;
-                if (!this._addedGeom.has(geom)) {
-                    this._editedGeom.add(geom);
+            // Reassign track in track map
+            let track = this._trackMap.get(trackId);
+            this._trackMap.set(newTrackId, track);
+            this._trackMap.delete(trackId);
+
+            // Update all geometries
+            for (let [ts0, geomSet] of this._frameMap) {
+                let geom = geomSet.get(trackId);
+                if (geom) {
+                    geom.id1 = newTrackId;
+                    geomSet.delete(trackId);
+                    geomSet.set(newTrackId, geom);
+
+                    if (!this._addedGeom.has(geom.id0)) {
+                        this._editedGeom.set(geom.id0, geom);
+                    }
                 }
-            });
+            };
+        }
+        return this.copy();
+    }
+
+    //-------------------------------------------------------------------------
+    removeTrack(trackId) {
+        for (let [ts0, geomSet] of this._frameMap) {
+            if (geomSet.has(trackId)) {
+                let state = {id1: trackId, ts0};
+                this._remove(frame, trackId, this._stateMap.get(state));
+            }
         }
         return this.copy();
     }
 
+    //-------------------------------------------------------------------------
     getFrame(frame) {
-        return this._frameMap.get(frame);
+        return Array.from(this._frameMap.get(frame).values());
     }
 
+    //-------------------------------------------------------------------------
     get length() {
         return this._frameMap.size;
     }
 
+    //-------------------------------------------------------------------------
     change(frame, trackId, g0, itemId = null) {
-        var geomSet = this.getFrame(frame);
-        var geomToChange = Array.from(geomSet).find((geom) => {
-            return geom.id1 === trackId && geom.ts0 === frame;
-        });
-        if (geomToChange) {
-            Object.assign(geomToChange, {
-                g0
-            });
-            if (!this._addedGeom.has(geomToChange)) {
-                this._editedGeom.add(geomToChange);
+        // Look up ID of possibly existing geometry
+        let state = {id1: trackId, ts0: frame};
+        let id0 = this._stateMap.get(state);
```

Comment:
You probably still working on this but I would simply point out that this _stateMap is not defined at this moment.

Reply:
Oops, yes... I was going back and forth between `_stateMap` and `_geomMap`, and some spots still had `_stateMap`. Fixed.